src/main/java/org/jenkinsci/plugins/parallel_test_executor/TestMode.java

```diff
@@ -0,0 +1,17 @@
+package org.jenkinsci.plugins.parallel_test_executor;
+
+public enum TestMode  {
+    JAVA ("Parallelize on test classes (Java)"),
+    TESTCASENAME ("Parallelize on test case name (Generic)"),
+    CLASSANDTESTCASENAME ("Parallelize on class and test case name (Generic)");

```

Comment:
Seems like this is on the right track, but should be refactored into an `AbstractDescribableImpl` whose `Descriptor`s with `@Symbol`s are an `ExtensionPoint`, so that support for Java is not baked into anything except the Java implementation, and support for other languages could be added just by introducing new modes, potentially from other plugins.

IIUC the critical bit would be something like

```java
public abstract Map<String, TestEntity> getData(ClassResult result);
```